pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -134,7 +135,7 @@ func (m *monitor) Run() {
 
 type monitors map[schema.GroupVersionResource]*monitor
 
-func (gb *GraphBuilder) controllerFor(logger klog.Logger, resource schema.GroupVersionResource, kind schema.GroupVersionKind) (cache.Controller, cache.Store, error) {
+func (gb *GraphBuilder) controllerFor(logger klog.Logger, resource schema.GroupVersionResource, kind schema.GroupVersionKind) (cache.Controller, cache.Store, func(), error) {
```

Comment:
I realized there's an existing bug in GC we need to fix before we merge this. If there's a discovery error today, GC drops the monitors for the resources within the group/versions that hit discovery errors. Then when discovery recovers, it readds those monitors, and now has duplicate event handlers.

While we're in a state with discovery errors, we don't actually want to drop existing synced informers for things in that group or we'll miss deletion information when the informer recovers. Fix in https://github.com/kubernetes/kubernetes/pull/117992